Src/Base/AMReX_VisMF.cpp

```diff
@@ -2217,6 +2245,248 @@ void VisMF::CloseAllStreams() {
+    const FABio& fio = FArrayBox::getFABio();
+    for (MFIter mfi(mf); mfi.isValid(); ++mfi)
+    {
+        std::memcpy(pld, &total_bytes, sizeof(int64_t));
+        pld += sizeof(int64_t);
+
+        const FArrayBox& fab = mf[mfi];
+
+        std::stringstream hss;
+        fio.write_header(hss, fab, ncomp);
+        total_bytes += static_cast<std::streamoff>(hss.tellp());
+        total_bytes += fab.size() * whichRD.numBytes();
+
+        // compute min and max
+        const Box& bx = mfi.validbox();
+
+        Real cmin, cmax;
+        for (int icomp = 0; icomp < ncomp; ++icomp) {
+            if (run_on_device) {
+                cmin = fab.min<RunOn::Device>(bx,icomp);
+                cmax = fab.max<RunOn::Device>(bx,icomp);
+            } else {
+                cmin = fab.min<RunOn::Host>(bx,icomp);
+                cmax = fab.max<RunOn::Host>(bx,icomp);
+            }
+
+            std::memcpy(pld, &cmin, sizeof(Real));
+            pld += sizeof(Real);
+            std::memcpy(pld, &cmax, sizeof(Real));
+            pld += sizeof(Real);
+        }
+    }
+    localdata[0] = total_bytes;
+
+    auto globaldata = std::make_shared<Vector<int64_t> >();
+    if (nprocs == 1) {
+        *globaldata = std::move(localdata);
+    }
+#ifdef BL_USE_MPI
+    else {
+        const Long n_global_nums = n_fab_nums * n_global_fabs + nprocs;
+        Vector<int> rcnt, rdsp;
+        if (myproc == io_proc) {
+            globaldata->resize(n_global_nums);
+            rcnt.resize(nprocs,1);
+            rdsp.resize(nprocs,0);
+            for (int k = 0; k < n_global_fabs; ++k) {
+                int rank = dm[k];
+                rcnt[rank] += n_fab_nums;
+            }
+            std::partial_sum(rcnt.begin(), rcnt.end()-1, rdsp.begin()+1);
+        } else {
+            globaldata->resize(1,0);
+            rcnt.resize(1,0);
+            rdsp.resize(1,0);
+        }
+        BL_MPI_REQUIRE(MPI_Gatherv(localdata.data(), localdata.size(), MPI_INT64_T,
+                                   globaldata->data(), rcnt.data(), rdsp.data(), MPI_INT64_T,
+                                   io_proc, ParallelDescriptor::Communicator()));
+    }
+#endif
+
+    auto myfabs = std::make_shared<Vector<FArrayBox> >();
+    for (MFIter mfi(mf); mfi.isValid(); ++mfi) {
+#ifdef AMREX_USE_GPU
+        if (data_on_device) {
+            myfabs->emplace_back(mf[mfi].box(), mf.nComp(), The_Pinned_Arena());
+            auto& new_fab = myfabs->back();
+            Gpu::dtoh_memcpy(new_fab.dataPtr(), mf[mfi].dataPtr(), new_fab.size()*sizeof(Real));
+        } else
+#endif
+        if (is_rvalue) {
+            myfabs->emplace_back(std::move(const_cast<FArrayBox&>(mf[mfi])));
+        } else {
+            myfabs->emplace_back(mf[mfi].box(), mf.nComp(), The_Cpu_Arena());
+            auto& new_fab = myfabs->back();
+            new_fab.copy<RunOn::Host>(mf[mfi]);
+        }
+    }
+
+    std::shared_ptr<FABio> fabio(new FABio_binary(FPC::NativeRealDescriptor().clone()));
```

Comment:
What keep this from going out of scope when the async call finishes? Is it that the shared pointer gets captured in the lamba, so it doesn't get destroyed until the lambda call finishes, which would be when the IO thread is done writing?

Reply:
Yes, exactly.  The `shared_ptr` is captured by value by the lambda function provided to `AsyncOut::Submit`.  The local copy gets destroyed, but the copy in the lambda function (wrapper in `std::function` for type erasure) is pushed on to a `queue` by the main thread.  When the IO thread wakes up, it will pop the function from the queue and do work.  After the work is one, it's destroyed.